lib/spack/spack/ci.py

```diff
@@ -2020,50 +2035,66 @@ def process_command(name, commands, repro_dir, run=True, exit_on_failure=True):
 
     Returns: the exit code from processing the command
     """
-    tty.debug("spack {0} arguments: {1}".format(name, commands))
 
+    tty.debug(f"spack {name} arguments: {commands}")
     if len(commands) == 0 or isinstance(commands[0], str):
         commands = [commands]
-
-    # Create a string [command 1] && [command 2] && ... && [command n] with commands
-    # quoted using double quotes.
-    args_to_string = lambda args: " ".join('"{}"'.format(arg) for arg in args)
-    full_command = " \n ".join(map(args_to_string, commands))
-
-    # Write the command to a shell script
-    script = "{0}.sh".format(name)
+    ext = "ps1" if IS_WINDOWS else "sh"
+    verbose_commands = "Set-PSDebug -Trace 2\n" if IS_WINDOWS else "set -x\n"
+    shebang = "" if IS_WINDOWS else "#!/bin/sh\n\n"
+    exit_error = '$ErrorActionPreference = "Stop"\n' if IS_WINDOWS else "set -e\n"
+    # ErrorActionPreference will handle PWSH commandlets (Spack calls),
+    # but we need to handle EXEs (git, etc) ourselves
+    catch_exe_failure = (
+        """
+if ($LASTEXITCODE -ne 0){
+    throw "Command {} has failed"
+}
+"""
+        if IS_WINDOWS
+        else ""
+    )
```

Comment:
(question) If we're doing this after each command in a powershell script, is `$ErrorActionPreference = "Stop"` redundant? I would only be right about this if failed PWSH commandlets have a nonzero exit code.

Reply:
Powershell commandlets have "non-terminating errors" (i.e. if you manually write to stderr w\ `Write-Error` during a script run). Some will terminate in this way even when they've failed, but their exit code is still 0. Both cases are required.